rfcs/text/0007_lifecycle_unblocked.md

```diff
@@ -0,0 +1,290 @@
+2. Synchronous context provider functions
+3. Core should not expose API's as observables
+
+# Motivation
+Plugin lifecycle methods and context provider functions are async
+(promise-returning) functions. Core runs these functions in series and waits
+for each plugin's lifecycle function to resolve before calling the next. This
+allows plugins to depend on the API's returned from other plugins.
+
+With the current design, a single lifecycle method or context provider that
+blocks will block all of Kibana from starting up.
+
+We should make it impossible for a single plugin lifecycle function to
+stall all of kibana.
+
+# Detailed design
+
+### 1. Synchronous lifecycle methods
+Lifecycle methods are synchronous functions, they can perform asynchronous
+operations but Core doesn't wait for these to complete. This guarantees that
+no plugin lifecycle function can block other plugins or core from starting up.
+
+Core will still expose special API's that are able block the setup lifecycle
+such as registering Saved Object migrations, but this will be limited to
+operations where the risk of blocking all of kibana starting up is limited.
+
+### 2. Synchronous Context Provider functions
+Making context provider functions synchronous guarantees that a context
+handler will never be blocked by registered context providers. They can expose
+asynchronous API's which could potentially have blocking behaviour. 
+
+```ts
+export type IContextProvider<
+  TContext extends Record<string, any>,
+  TContextName extends keyof TContext,
+  TProviderParameters extends any[] = []
+> = (
+  context: Partial<TContext>,
+  ...rest: TProviderParameters
+) => TContext[TContextName];
+```
+
+### 3. Core should not expose API's as observables
+All Core API's should be reactive, when internal state changes their behaviour
+should change accordingly. But, exposing these internal state changes as part
+of the API contract leaks internal implementation details consumers can't do
+anything useful with and don't care about.
+
+For example: Core currently exposes `core.elasticsearch.adminClient$`, an
+Observable which emits a pre-configured elasticsearch client every time there's
+a configuration change. This includes changes such as the elasticsearch
+cluster `hosts` that alter the behaviour of the elasticsearch client. As a
+plugin author who wants to make search requests against elasticsearch I
+shouldn't have to care about, react to, or keep track of, how many times the
+underlying configuration has changed. I want to use the `callAsInternalUser`
+method and I expect Core to use the most up to date configuration to send this
+request to the correct `hosts`.
+
+This does not mean we should remove all observables from Core's API's. When an
+API consumer is interested in the *state changes itself* it absolutely makes
+sense to expose this as an Observable. Good examples of this is exposing
+plugin config as this is state that changes over time to which a plugin should
```

Comment:
I didn't say we should **remove** observables. I proposed to expose a version that supports sync access. I see it's a valid use case when plugins read config during setup. Isn't it? 
Also it will require some refactoring for the existing code https://github.com/elastic/kibana/blob/9d69b72a5f200e58220231035b19da852fc6b0a5/src/plugins/timelion/server/plugin.ts#L40

Reply:
> I see it's a valid use case when plugins read config during setup. Isn't it?

Sorry my comment was sparked by your question, but transformed into a more general comment on the RFC. The point I'm trying to make is that I don't think we **need** to remove the Observable APIs from the ElasticsearchService as proposed in the Adoption strategy. I don't see a reason that this is actually necessary in that case.

Config may be a valid thing that should expose a sync version during `setup` so that plugins can conditionally register items based on the config.

With Elasticsearch clients, the APIs of the clients themselves are already async, so making the API for getting a client instance being sync is not a blocker for making this RFC viable.

```ts
class Plugin {
  setup(core) {
    return {
      async someQuery() {
        // Whether or not this async does not change the signature of the API
        const client = core.elasticsearch.dataClient$.pipe(take(1)).toPromise();
        return client.callWithInternalUser('search', ...);
      }
    }
  }
}
```

I still think making these sync is a good idea, I just don't think in the case of ElasticsearchService that this is necessary to accomplish the goals of this RFC.

Reply:
I haven't double checked this to be 100% sure it works in all cases, but from running a quick test, it seems like config is synchronously available:

```
console.log('about to subscribe');
configService.atPath('key').subscribe(value => {
  console.log('got value');
});
console.log('after subscribe');
```

We could provide a convenience property like `BehaviourSubject.latest` to make this more obvious.

But I think I agree with Josh that it wouldn't be required but will make the ergonomics easier.

Reply:
I went through all async NP / Legacy plugins and as far as I can tell they're only async because it improves the ergonomics of consuming Core or dependency API's (like `elasticsearch.adminClient$`, `config$`, `createCookieSessionStorageFactory(): Promise< SessionStorageFactory>`. Another way to look at it is that `async setup()` is mainly used to unwrap values from Observables.

How we expose config values (and other values that could change over time) can greatly affect the architecture of plugins:
1. async lifecycles + observable config: plugins choose
2. sync lifecycles + sync config (observable optional): plugins choose
3. sync lifecycles + observable: plugins must be reactive

When plugins choose, there's a strong bias towards maintaining the status quo which is not to be reactive.

If Core wants to promote a more reactive Kibana, then simply unwrapping these values in `async setup()` or providing sync accessors is an anti-pattern. However, requiring plugins to refactor their code to be completely reactive is a lot of additional work.

Reply:
> 3\. sync lifecycles + observable: plugins must be reactive

Is this true though? In the example in this RFC [here](https://github.com/elastic/kibana/pull/45796/files#diff-f07024532a7a5d4c1bfc812383d6542dR170) the plugin only calculates the config once, and then reuses the Promise in each API call.

It's true that example could be changed so that the config is calculated each time, but I do think the synchronous lifecycle forces this pattern.

Reply:
> If Core wants to promote a more reactive Kibana, then simply unwrapping these values in async setup() or providing sync accessors is an anti-pattern. However, requiring plugins to refactor their code to be completely reactive is a lot of additional work.

Should the core control how a plugin consumes a config value? A plugin knows better if a config value can be updated in runtime.
If a value **should** be updated without restart, it will be a business requirement. 
If a value cannot be updated without restart, it's **required** to be consumed only once and in case of updates to warn a user that new config value cannot be applied without a restart.
If business changes requirements and a plugin should support config runtime updates, it will adopt the code and a public contract eventually.